File: daily_python/3721_Longest_Balanced_Subarray_II.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

struct LazyTag{
    int toAdd = 0;

    LazyTag() {};
    LazyTag(int x) : toAdd(x) {};

    bool hasTag() const { return toAdd != 0; }
    void clear(){ toAdd = 0; }

    LazyTag& operator +=(const LazyTag& other){
        this->toAdd += other.toAdd;
        return *this;
    }
};

struct SegmentTreeNode{
    int minValue;
    int maxValue;
    LazyTag lazyTag;
    SegmentTreeNode() {};
    SegmentTreeNode(int x) : minValue(x), maxValue(x) {};
};

class SegmentTree{
public:
    int n;
    vector<SegmentTreeNode> tree;
    SegmentTree(const vector<int>& nodes){
        n = nodes.size();
        tree.resize(4 * n + 1);
        buildTree(nodes, 1, n, 1);
    }

    void periodAdd(int start, int end, int val){
        LazyTag t(val);
        periodUpdate(start, end, t, 1, n, 1);
    }

    int findLastZero(int start){
        if (start > n) return -1;
        return find(start, n, 0, 1, n, 1);
    }
private:
    void buildTree(const vector<int>& nodes, int start, int end, int nodeIdx){
        if (start == end){
            tree[nodeIdx] = SegmentTreeNode(nodes[start - 1]);
            return;
        }
        int mid = start + (end - start) / 2;
        buildTree(nodes, start, mid, nodeIdx * 2);
        buildTree(nodes, mid + 1, end, nodeIdx * 2 + 1);
        pushup(nodeIdx);
    }
    void pushup(int nodeIdx){
        tree[nodeIdx].minValue = min(tree[nodeIdx * 2].minValue, tree[nodeIdx * 2 + 1].minValue);
        tree[nodeIdx].maxValue = max(tree[nodeIdx * 2].maxValue, tree[nodeIdx * 2 + 1].maxValue);
    }
    void pushdown(int nodeIdx){
        // refresh all node under current node
        if (tree[nodeIdx].lazyTag.hasTag()){
            LazyTag tag = tree[nodeIdx].lazyTag;
            applyTag(tag, nodeIdx * 2);
            applyTag(tag, nodeIdx * 2 + 1);
            tree[nodeIdx].lazyTag.clear();
        }
    }
    void applyTag(const LazyTag& t, int nodeIdx){
        tree[nodeIdx].minValue += t.toAdd;
        tree[nodeIdx].maxValue += t.toAdd;
        tree[nodeIdx].lazyTag += t;
    }
    void periodUpdate(int start, int end, const LazyTag& t, int currNodeLeft, int currNodeRight, int nodeIdx){
        if (start <= currNodeLeft && currNodeRight <= end){
            applyTag(t, nodeIdx);
            return;
        }
        pushdown(nodeIdx);
        int mid = currNodeLeft + (currNodeRight - currNodeLeft) / 2;
        if (start <= mid) periodUpdate(start, end, t, currNodeLeft, mid, nodeIdx * 2);
        if (end >= mid + 1) periodUpdate(start, end, t, mid + 1, currNodeRight, nodeIdx * 2 + 1);
        pushup(nodeIdx);
    }
    int find(int start, int end, int val, int currNodeLeft, int currNodeRight, int nodeIdx){
        // not find val (any interval in which max and min contain val is guaranteed to contain val in this interval)
        if (tree[nodeIdx].minValue > val || tree[nodeIdx].maxValue < val) return -1;
        if (currNodeLeft == currNodeRight) return currNodeLeft;
        pushdown(nodeIdx);
        int mid = currNodeLeft + (currNodeRight - currNodeLeft) / 2;
        if (end >= mid + 1){
            int res = find(start, end, val, mid + 1, currNodeRight, nodeIdx * 2 + 1);
            if (res != -1) return res;
        }
        if (start <= mid && currNodeLeft <= end){
            return find(start, end, val, currNodeLeft, mid, nodeIdx * 2);
        }
        return -1;
    }
};

class Solution {
public:
    int longestBalanced(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, queue<int>> m;
        auto sgn = [](int x){ return (x % 2) == 0? 1 : -1; };
        vector<int> prefixSum(n, 0);
        prefixSum[0] = sgn(nums[0]);
        m[nums[0]].push(1);
        for (int i = 1; i < n; i++){
            prefixSum[i] = prefixSum[i-1];
            auto& latest = m[nums[i]];
            if (latest.empty()){
                prefixSum[i] += sgn(nums[i]);
            }
            latest.push(i + 1);
        }
        SegmentTree seg(prefixSum);
        int ans = 0;
        for (int i = 0; i < n; i++){
            ans = max(ans, seg.findLastZero(i + 1 + ans) - i);
            int nextPos = n + 1;
            m[nums[i]].pop();
            // idx in map already convert to tree idx(+1)
            if (!m[nums[i]].empty()) nextPos = m[nums[i]].front();
            seg.periodAdd(i + 1, nextPos - 1, -sgn(nums[i]));
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {2,5,4,3};
    // vector<int> nums = {3,2,2,5,4};
    // vector<int> nums = {1,2,3,2};
    Solution S;
    cout << S.longestBalanced(nums) << endl;
    return 0;
}

File: daily_python/3721_Longest_Balanced_Subarray_II.py
import os
import sys
from typing import List
from collections import defaultdict, deque

class SegmentTree:
    def __init__(self, prefix: List[int]):
        self.n = len(prefix)
        self.minv = [0] * (4 * self.n + 1)
        self.maxv = [0] * (4 * self.n + 1)
        self.lazy = [0] * (4 * self.n + 1)
        self._build(prefix, 1, self.n, 1)
    def _build(self, prefix: List[int], l: int, r: int, nodeidx: int):
        if l == r:
            val = prefix[l - 1]
            self.minv[nodeidx] = val
            self.maxv[nodeidx] = val
            return
        mid = (l + r) // 2
        self._build(prefix, l, mid, nodeidx * 2)
        self._build(prefix, mid + 1, r, nodeidx * 2 + 1)
        self._pushup(nodeidx)

    def _pushup(self, nodeidx: int):
        self.minv[nodeidx] = min(self.minv[nodeidx * 2], self.minv[nodeidx * 2 + 1])
        self.maxv[nodeidx] = max(self.maxv[nodeidx * 2], self.maxv[nodeidx * 2 + 1])

    def _pushdown(self, nodeidx: int):
        if self.lazy[nodeidx] != 0:
            v = self.lazy[nodeidx]
            self._applyTag(v, nodeidx * 2)
            self._applyTag(v, nodeidx * 2 + 1)
            self.lazy[nodeidx] = 0
    
    def _applyTag(self, t, nodeidx):
        self.minv[nodeidx] += t
        self.maxv[nodeidx] += t
        self.lazy[nodeidx] += t
    
    def _periodUpdate(self, start: int, end: int, t: int, l: int, r: int, nodeidx: int):
        if start <= l and end >= r:
            self._applyTag(t, nodeidx)
            return
        self._pushdown(nodeidx)
        mid = (l + r) // 2
        if start <= mid:
            self._periodUpdate(start, end, t, l, mid, nodeidx * 2)
        if end >= mid + 1:
            self._periodUpdate(start, end, t, mid + 1, r, nodeidx * 2 + 1)
        self._pushup(nodeidx)
    
    def _find(self, start: int, end: int, val: int, l: int, r: int, nodeidx: int) -> int:
        if self.minv[nodeidx] > val or self.maxv[nodeidx] < val:
            return -1
        if l == r:
            return l
        self._pushdown(nodeidx)
        mid = (l + r) // 2
        if end >= mid + 1:
            res = self._find(start, end, val, mid + 1, r, nodeidx * 2 + 1)
            if res != -1:
                return res
        if start <= mid and end >= l:
            return self._find(start, end, val, l, mid, nodeidx * 2)
        return -1
    
    def periodAdd(self, start: int, end: int, t: int):
        self._periodUpdate(start, end, t, 1, self.n, 1)

    def findZero(self, start: int) -> int:
        return self._find(start, self.n, 0, 1, self.n, 1)

class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        n = len(nums)
        def sgn(x : int) -> int:
            return 1 if x % 2 == 0 else -1
        prefix = [0] * n
        prefix[0] = sgn(nums[0])
        occ = defaultdict(deque)
        occ[nums[0]].append(1)
        for i in range(1, n):
            prefix[i] = prefix[i - 1]
            if not occ[nums[i]]:
                prefix[i] += sgn(nums[i])
            occ[nums[i]].append(i + 1)

        seg = SegmentTree(prefix)        
        ans = 0
        for i in range(n):
            ans = max(ans, seg.findZero(i + 1 + ans) - i)
            occ[nums[i]].popleft()
            next_pos = seg.n + 1 if not occ[nums[i]] else occ[nums[i]][0]
            seg.periodAdd(i + 1, next_pos - 1, -sgn(nums[i]))
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.longestBalanced(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')